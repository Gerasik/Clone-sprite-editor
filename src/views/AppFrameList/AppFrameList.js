export default class AppFrameList {
  constructor(className, parentBlock, size) {
    this.frames = [];
    this.className = className;
    this.parentBlock = parentBlock;
    this.block = null;
    this.current = null;
    this.size = size;
  }

  render() {
    const frameList = document.createElement('section');
    frameList.classList.add(this.className);
    frameList.innerHTML = `<div class="${this.className}__add-frame">
                              <span>&#xf067</span>
                              <span>Add new frame</span>
                           <div>`;
    this.parentBlock.appendChild(frameList);
    this.block = frameList;
    if (this.frames.length === 0) {
      this.createFrame();
    }
  }

  createFrame(prewFrame) {
    const newFrame = document.createElement('div');
    const canvas = document.createElement('canvas');
    const number = document.createElement('span');
    const remove = document.createElement('span');
    const copy = document.createElement('span');
    const move = document.createElement('span');
    canvas.setAttribute('width', this.size.width);
    canvas.setAttribute('height', this.size.width);
    number.classList.add(`${this.className}__number`);
    remove.classList.add(`${this.className}__remove`);
    copy.classList.add(`${this.className}__copy`);
    move.classList.add(`${this.className}__move`);
    number.innerHTML = this.block.childElementCount;
    remove.innerHTML = '&#xf2ed';
    copy.innerHTML = '&#xf0c5';
    move.innerHTML = '&#xf0b2';
    newFrame.classList.add(`${this.className}__frame`);
    newFrame.appendChild(number);
    newFrame.appendChild(remove);
    newFrame.appendChild(copy);
    newFrame.appendChild(move);
    newFrame.appendChild(canvas);
    newFrame.setAttribute('draggable', true);
    if (prewFrame) {
      prewFrame.parentElement.insertBefore(newFrame, prewFrame.nextSibling);
    } else {
      this.block.insertBefore(newFrame, this.block.lastChild);
    }
    this.toggleCurrFrame(newFrame);
    return newFrame;
  }

  toggleCurrFrame(frame) {
    if (this.current) this.current.classList.remove('frame-list__curr-frame');
    frame.classList.add(`${this.className}__curr-frame`);
    this.current = frame;
  }

  createCopyFrame(prevFrame) {
    const newFrame = this.createFrame(prevFrame);
    const newCanvas = newFrame.getElementsByTagName('canvas')[0];
    const oldCanvas = prevFrame.getElementsByTagName('canvas')[0];
    const data = oldCanvas.getContext('2d').getImageData(0, 0, oldCanvas.width, oldCanvas.height);
    newCanvas.getContext('2d').putImageData(data, 0, 0);
    this.toggleCurrFrame(newFrame);
    return newFrame;
  }

  updateCount() {
    const elements = this.block.children;
    for (let i = 1; i < elements.length; i += 1) {
      const number = elements[i - 1].querySelector('.frame-list__number');
      number.innerHTML = i;
    }
  }
}
