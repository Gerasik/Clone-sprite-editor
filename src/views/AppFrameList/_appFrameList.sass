$main-color: #000
$hover-main-color: #ff0
.frame-list
  align-self: flex-start
  padding: 10px 0
  font-size: 14px
  overflow-y: auto
  max-height: 90%
  position: relative
  &__frame,
  &__add-frame
    width: 80px
    height: 80px
    border: 3px solid $main-color
    border-radius: 5px
    display: flex
    justify-content: center
    align-items: center
    transition: all 0.2s ease-out
    &.frame-list__curr-frame,
    &:hover
      border-color: $hover-main-color
      span:nth-child(1)
        color: $main-color
      .frame-list__remove,
      .frame-list__copy,
      .frame-list__move
        display: flex
      .frame-list__number
        background-color: $hover-main-color
    &.frame-list__curr-frame
      span:nth-child(1)
        color: $main-color
        background-color: $hover-main-color
  &__add-frame
    cursor: pointer
    span:nth-child(1)
      font-family: 'Awesome'
      font-size: 24px
      margin: auto 5px auto 5px
      transition: all 0.2s ease-out
  &__frame
    position: relative
    &.over
      border: 3px dashed $hover-main-color
      background-color: red
    canvas
      width: 100%
      height: 100%
      image-rendering: pixelated
      pointer-events: none
  &__number,
  &__remove,
  &__copy,
  &__move
    position: absolute
    width: 20px
    height: 20px
    display: flex
    justify-content: center
    align-items: center
  &__remove,
  &__copy,
  &__move
    font-family: 'Awesome'
    cursor: pointer
    background-color: $hover-main-color
    display: none
  &__number
    top: 0
    left: 0
    border-bottom-right-radius: 5px
    background-color: $main-color
    color: #fff
  &__remove
    top: 0
    right: 0
    border-bottom-left-radius: 5px
  &__copy
    left: 0
    bottom: 0
    border-top-right-radius: 5px
  &__move
    bottom: 0
    right: 0
    border-top-right-radius: 5px
