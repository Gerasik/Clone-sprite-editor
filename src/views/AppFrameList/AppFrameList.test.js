import AppFrameList from './AppFrameList';
import 'jest-canvas-mock';

const cont = {
  frames: [],
  className: 'frame-list',
  parentBlock: document.body,
  size: { width: 10 },
  createFrame: AppFrameList.prototype.createFrame,
  toggleCurrFrame: AppFrameList.prototype.toggleCurrFrame,
  block: document.body,
  current: null,
};

describe('new AppFrameList', () => {
  it('Should be an instance of Object', () => {
    expect(new AppFrameList('header', document.body, 25)).toBeInstanceOf(Object);
  });
  it('Should be contain right parameters', () => {
    const newObj = new AppFrameList('header', document.body, 25);
    expect(newObj.className).toBe('header');
    expect(newObj.block).toBe(null);
    expect(newObj.frames).toBeInstanceOf(Array);
    expect(newObj.parentBlock).toBe(document.body);
    expect(newObj.current).toBe(null);
    expect(newObj.size).toBe(25);
  });
});

describe('AppFrameList.prototype.toggleCurrFrame', () => {
  const frame = document.createElement('div');
  const func = AppFrameList.prototype.toggleCurrFrame.bind(cont, frame);
  func();
  it('Should be an instance of Function', () => {
    expect(func).toBeInstanceOf(Object);
  });
  it('Should be create right class', () => {
    expect(cont.current.classList[0]).toBe(`${cont.className}__frame`);
  });
});

describe('AppFrameList.prototype.render', () => {
  it('Should be render slide', () => {
    AppFrameList.prototype.render.bind(cont)();
    expect(document.body.innerHTML).toMatchSnapshot();
  });
  cont.frames = [1];
  it('Should be create only 7 child', () => {
    AppFrameList.prototype.render.bind(cont)();
    expect(cont.parentBlock.children.length).toBe(7);
  });
});


describe('AppFrameList.prototype.createFrame', () => {
  const ans = AppFrameList.prototype.createFrame.bind(cont, false)();
  it('Should be consist canvas', () => {
    expect(ans.getElementsByTagName('canvas')[0].tagName).toBe('CANVAS');
  });
  it('Should be consist 5 items', () => {
    expect(ans.children.length).toBe(5);
  });
});

describe('AppFrameList.prototype.updateCount', () => {
  const bodyBlock = document.body;
  bodyBlock.innerHTML = '<div><span class="frame-list__number">3</span></div><div><span class="frame-list__number">1</span></div><div><span class="frame-list__number">2</span></div>';
  const context = {
    block: bodyBlock,
  };
  AppFrameList.prototype.updateCount.bind(context, false)();
  const list = context.block.querySelectorAll('.frame-list__number');
  it('Should be set correct order', () => {
    expect(list[0].innerHTML).toBe('1');
    expect(list[1].innerHTML).toBe('2');
  });
});

describe('AppFrameList.prototype.createCopyFrame', () => {
  const newFrame = document.createElement('div');
  const canvas = document.createElement('canvas');
  canvas.setAttribute('width', 5);
  canvas.setAttribute('height', 5);
  newFrame.appendChild(canvas);
  cont.parentBlock.appendChild(newFrame);
  const elem = AppFrameList.prototype.createCopyFrame.bind(cont, newFrame)();
  it('Should be set correct order', () => {
    expect(elem.classList[0]).toBe('frame-list__frame');
  });
});
