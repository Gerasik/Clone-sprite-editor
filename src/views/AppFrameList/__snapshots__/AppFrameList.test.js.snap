// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppFrameList.prototype.render Should be render slide 1`] = `
"<div><span class=\\"frame-list__number\\">1</span></div><div><span class=\\"frame-list__number\\">2</span></div><div><span class=\\"frame-list__number\\">2</span></div><div><canvas width=\\"5\\" height=\\"5\\"></canvas></div><div class=\\"frame-list__frame frame-list__curr-frame\\" draggable=\\"true\\"><span class=\\"frame-list__number\\">4</span><span class=\\"frame-list__remove\\"></span><span class=\\"frame-list__copy\\"></span><span class=\\"frame-list__move\\"></span><canvas width=\\"10\\" height=\\"10\\"></canvas></div><section class=\\"frame-list\\"><div class=\\"frame-list__add-frame\\">
                              <span></span>
                              <span>Add new frame</span>
                           <div></div></div></section>"
`;
